lib/tables.c

```diff
@@ -4675,3 +4675,146 @@ table_collection_simplify(table_collection_t *self,
     simplifier_free(&simplifier);
     return ret;
 }
+
+
+/*******************
+ * Cleaning tables *
+ ******************/
+
+typedef struct {
+    site_table_t *sites;
+    mutation_table_t *mutations;
+    /* Map of old site IDs to new site IDs. */
+    site_id_t *site_id_map;
+} table_cleaner_t;
+
+static int table_cleaner_run(table_cleaner_t *self)
+{
+    // Remove any sites with duplicate positions, retaining only the *first*
+    // one. Assumes the tables have been sorted, throwing an error if not.
+    // TODO: this could be done in place rather than by constructing a new
+    // table.
+
+    int ret = 0;
+
+    if (self->sites->num_rows <= 1) {
+        // nothing to do
+        goto out;
+    }
+
+    table_size_t j, k, site_j;
```

Comment:
Style-wise, I try to keep all variable declarations before any code. I know this isn't necessary in C99, but I think it's a good habit and encourages short functions. Also, it's important safety wise. Suppose you had
```c
int function(int x) 
{
    int ret = 0;
    if (x < 0) {
        ret = -1;   
        goto out;
     }
     int *p = malloc(x * sizeof(int));
     ..
out:
     if (p != NULL) {
         free(p);  
     }
     return ret;
}
```
In this case, the compiler will probably be smart enough to know that ``p`` is undefined in the error case, but there will be situations where it can't. So, to guarantee safety, I always make sure that every variable is declared and all pointers have been assigned some value **before** running any code.